### PR TITLE
Profiling prints for tf.keras.

### PR DESCRIPTION
**Pull request by bfrieds**
Tested by running the tf keras mnist example with the following setups (all with aggregation frequency of 2):

profile freq of 0: nothing :)
profile freq of 1: got fast, slow, fast, slow
profile freq of 2: got slow, slow, slow
profile freq of 3: got fast, slow, fast, slow
profile freq of 8: got slow, slow, slow

### CODE REVIEW

File: horovod/_keras/__init__.py [hunk 1]
```diff
@@ -167,32 +185,83 @@ def allreduce_grads():
...
+
+            def profile_start():
+                if self._profile_frequency == 0:
+                    return tf.no_op()
+                return tf.cond(tf.equal(self.profile_counter, self._profile_frequency - 1), log_comm_start, tf.no_op)
+
+            def log_comm_start():
+                profile_base_info = trim_last_curly_brace(get_profile_info("communication", "B"))
+                self.start_timestamp = tf.timestamp() * 1e6
+                return tf.print(profile_base_info, ", \"ts\": ", self.start_timestamp, "}", sep="", output_stream="file:///profile/pedl_profiling.log")
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: Should we make the `output_stream` file path configureable?

File: horovod/_keras/__init__.py [hunk 2]
```diff
@@ -167,32 +185,83 @@ def allreduce_grads():
...
-                            tf.equal(self.counter, 0), clear_grads, tf.no_op)
-                        with tf.control_dependencies([clear_op]):
-                            aggregation_ops_list = aggregate_grads()
-
-                    aggregation_ops = tf.group(*aggregation_ops_list)
-                    with tf.control_dependencies([aggregation_ops]):
-                        update_ops = [self.counter.assign_add(tf.constant(1))]
-                else:
-                    update_ops = []
+                with tf.variable_scope("aggregation_variables", reuse=True):
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: Do we need this variable scope here? And if we do, we should probably not have nested calls to it (L241).

File: horovod/_keras/__init__.py [hunk 3]
```diff
@@ -167,32 +185,83 @@ def allreduce_grads():
...
+                return s[:-1]
+
+            def profile_start():
+                if self._profile_frequency == 0:
+                    return tf.no_op()
+                return tf.cond(tf.equal(self.profile_counter, self._profile_frequency - 1), log_comm_start, tf.no_op)
+
+            def log_comm_start():
+                profile_base_info = trim_last_curly_brace(get_profile_info("communication", "B"))
+                self.start_timestamp = tf.timestamp() * 1e6
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: ? add a comment for why the `1e6` is needed.

File: horovod/_keras/__init__.py [hunk 4]
```diff
@@ -167,32 +185,83 @@ def allreduce_grads():
...
+                    raise AssertionError(f"Expected last character in \"{s}\" to be \"}}\", but got \"{s[-1]}\"")
+                return s[:-1]
+
+            def profile_start():
+                if self._profile_frequency == 0:
+                    return tf.no_op()
+                return tf.cond(tf.equal(self.profile_counter, self._profile_frequency - 1), log_comm_start, tf.no_op)
+
+            def log_comm_start():
+                profile_base_info = trim_last_curly_brace(get_profile_info("communication", "B"))
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: Might be useful to explain this going the chrome formatting (hence the "B" and "E").

File: horovod/_keras/__init__.py (hunk 2 above)
Comment (bfrieds) [reply horovod/_keras/__init__.py]: Not positive if we do, but technically there are no nested calls to it, the scope is only used for line 238

File: horovod/_keras/__init__.py (hunk 1 above)
Comment (bfrieds) [reply horovod/_keras/__init__.py]: I have a few concerns with configuring this. Not configuring this allows us to easily build features around this. Would we want users to be able to configure different files for our Horovod stuff and PEDL stuff? We don't necessarily want to make everything configureable, and think it's fine to not configure this for now. If we want to make it configureable in the future it should not be bad.

**Comment by aaron276h** [comment]
@bfrieds It's up to you if you want leave the refactoring out the profiler until after `TFEstimator`. Very reasonable to wait and see what is the best way for them to "share" the code.

For the configuring the output file, we would definitely want to keep it consistent by passing the same file into the Horovod optimizer and the TFKeras callback. Making sure they explicitly consistent (rather than hardcoding the same thing in both places) is the main reason I think we should configure this. Possibly the right solution is to hardcode this in `TFKerasTrial.py` and then pass that around.

For the variable scope, you are right i misread the indentations. Still doesn't seem like we need the one on L237, but TF you never know.

**Comment by bfrieds** [comment]
> Making sure they explicitly consistent (rather than hardcoding the same thing in both places) is the main reason I think we should configure this. Possibly the right solution is to hardcode this in TFKerasTrial.py and then pass that around.

I see -- I can get behind this. That way if we want to make it configureable to end-users in the future, that just involves a change in one place.

File: horovod/_keras/__init__.py [hunk 5]
```diff
@@ -167,32 +188,97 @@ def allreduce_grads():
...
+                if not self._profile_frequency or not self._profile_filename:
+                    return tf.no_op()
+                return tf.cond(tf.equal(self.profile_counter, self._profile_frequency - 1), log_comm_start, tf.no_op)
+
+            def log_comm_start():
+                """
+                Log communication end profiling information.
+
+                Returns a tf.print operation that writes profiling information
+                for the end of communication to a file in the chrome://tracing
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: `for the start of communication`

File: horovod/_keras/__init__.py [hunk 6]
```diff
@@ -167,32 +188,97 @@ def allreduce_grads():
...
+            def log_comm_end():
+                """
+                Log communication end profiling information.
+
+                Returns a tf.print operation that writes profiling information
+                for the end of communication to a file in the chrome://tracing
+                format.
+                """
+                profile_base_info = trim_last_curly_brace(get_profile_info("communication", "E"))
+                # chrome//tracing uses milliseconds since epoch but timestamp is seconds since epoch. Multiply by 1e6 to get milliseconds.
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: nit: could we break L240 and L243 they are getting a big long.

File: horovod/_keras/__init__.py [hunk 7]
```diff
@@ -167,32 +188,97 @@ def allreduce_grads():
...
+                }
+                info.update(kwargs)
+                return json.dumps(info)
+
+            def profile_end():
+                if not self._profile_frequency or not self._profile_filename:
+                    return tf.no_op()
+                cond = tf.cond(tf.equal(self.profile_counter, self._profile_frequency - 1), log_comm_end, tf.no_op)
+                with tf.control_dependencies([cond]):
+                    return tf.cond(tf.equal(self.profile_counter, self._profile_frequency - 1), clear_profile_counter, increment_profile_counter)
```
Comment (aaron276h) [inline horovod/_keras/__init__.py]: ditto

**Pull request merged**